subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -126,21 +135,35 @@ public boolean isDependencyToAll(String className) {
     }
 
     /**
-     * Recursively accumulate dependent classes and resources.  Dependent classes discovered can themselves be used to query
-     * further dependents, while resources are just data accumulated along the way.
+     * Accumulate dependent classes and resources. Dependent classes discovered can themselves be used to query
+     * further dependents, while resources are just data accumulated along the way. Recurses for classes that
+     * are "publicly accessbile", i.e. classes that are not just used privately in a class.
      */
-    private void recurseDependentClasses(Set<String> visitedClasses, Set<String> resultClasses, Set<GeneratedResource> resultResources, Iterable<String> dependentClasses) {
-        for (String d : dependentClasses) {
+    private void processDependentClasses(Set<String> visitedClasses, Set<String> privateResultClasses, Set<String> accessibleResultClasses, Set<GeneratedResource> resultResources, Iterable<String> privateDependentClasses, Iterable<String> accessibleDependentClasses) {
+        for (String d : privateDependentClasses) {
             if (!visitedClasses.add(d)) {
                 continue;
             }
-            if (!isNestedClass(d)) {
-                resultClasses.add(d);
```

Comment:
@big-guy if you didn't already, can you clean up these inline comments? They are just confusing given that they are referring to some historical state of the code. In fact, it's probably best to remove all inline comments and add method-level/class-level docs instead where necessary.